### PR TITLE
AMD reordering and reorder heatmaps

### PR DESCRIPTION
**Pull request by AmroAlJundi**
Added AMD reorder:
* The reorder class.
* Modifications to the "Available" page.
* Added AMD to the optional dependencies github action.
* Added tests.

Added reorder heatmaps:
* Added the preprocess.
* Modification to the "Available" page.
* Added tests.

Additional changes:
* Changed `std::unique_ptr` constructors from default constructors to `make_unique` in preprocessor constructors.
* Removed adding converters to reordering in how-to guide.

### CODE REVIEW

File: src/sparsebase/preprocess/preprocess.cc [hunk 1]
```diff
@@ -415,6 +415,44 @@ IDType *RCMReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
 
+#ifdef USE_AMD_ORDER
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <amd.h>
+#ifdef __cplusplus
+}
+#endif
+
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:426]: A default constructor must be implemented. AMDReorderCSR function needs to be implemented in the default constructor as well.

File: src/sparsebase/preprocess/preprocess.cc [hunk 2]
```diff
@@ -415,6 +415,44 @@ IDType *RCMReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+  auto params = static_cast<AMDReorderParams*>(p);
+  long long n = csr->get_dimensions()[0];
+  long *xadj_long = csr_converted->get_row_ptr(), *adj_long = csr_converted->get_col();
+  long * i_order = new long[n];
+  double *Control = new double[AMD_CONTROL];
+  Control[AMD_DENSE] = params->dense;
+  Control[AMD_AGGRESSIVE] = params->aggressive ;
+  double *Info = nullptr; //new double[AMD_INFO]; // Auxiliary data
+  int status = amd_l_order(n, xadj_long, adj_long, i_order, Control, Info);
+  if (status != 0){
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:447]: it is better to throw a meaningful error here.

Comment (Atahanak) [reply src/sparsebase/preprocess/preprocess.cc:447]: also nice to extend our exceptions :).

File: src/sparsebase/preprocess/preprocess.cc [hunk 3]
```diff
@@ -1634,6 +1669,86 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
 
 #endif
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1671]: Default constructor is also needed for HeatMap.

File: src/sparsebase/preprocess/preprocess.cc [hunk 4]
```diff
@@ -415,6 +415,44 @@ IDType *RCMReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+AMDReorder<IDType, NNZType, ValueType>::AMDReorder(AMDReorderParams p){
+  this->params_ = std::make_unique<AMDReorderParams>(p);
+  this->RegisterFunction({format::CSR<IDType, NNZType, ValueType>::get_id_static()},
+                         AMDReorderCSR);
+}
+
+template <typename IDType, typename NNZType, typename ValueType>
+IDType* AMDReorder<IDType, NNZType, ValueType>::AMDReorderCSR(std::vector<format::Format*> formats, PreprocessParams* p){
+  auto csr = formats[0]->AsAbsolute<format::CSR<IDType, NNZType, ValueType>>();
+  auto csr_converted = csr->template Convert<format::CSR, long, long, ValueType>(false);
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:437]: a type exception needs to be thrown if the conversion fails.

File: src/sparsebase/preprocess/preprocess.cc [hunk 5]
```diff
@@ -1634,6 +1669,86 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+  auto n = csr->get_dimensions()[0];
+  auto row_ptr = csr->get_row_ptr();
+  auto adj = csr->get_col();
+  long long max_bw = 0;
+  double mean_bw = 0;
+
+  long long bsize = n / b;
+
+  // matrix of size num_parts x num_parts with number of edges in each square
+  auto density = new long long*[b];
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1705]: why do we use long long here?
num_parts can not be larger than NNZType?

File: src/sparsebase/preprocess/preprocess.cc [hunk 6]
```diff
@@ -1634,6 +1669,86 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+format::FormatOrderOne<FloatType>* ReorderHeatmap<IDType, NNZType, ValueType, FloatType>::ReorderHeatmapCSRandArray(std::vector<format::Format*> formats, PreprocessParams* poly_params){
+  auto csr = formats[0]->AsAbsolute<format::CSR<IDType, NNZType, ValueType>>();
+  auto order = formats[1]->AsAbsolute<format::Array<IDType>>()->get_vals();
+  auto* params = static_cast<ReorderHeatmapParams*>(poly_params);
+  int b = params->num_parts;
+  auto n = csr->get_dimensions()[0];
+  auto row_ptr = csr->get_row_ptr();
+  auto adj = csr->get_col();
+  long long max_bw = 0;
+  double mean_bw = 0;
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1700]: We must check if num parts is larger than number of rows

File: src/sparsebase/preprocess/preprocess.cc (hunk 4 above)
Comment (AmroAlJundi) [reply src/sparsebase/preprocess/preprocess.cc]: True.

**Review by Atahanak — CHANGES_REQUESTED** [review]
so minor changes are required.

File: src/sparsebase/preprocess/preprocess.cc (hunk 5 above)
Comment (AmroAlJundi) [reply src/sparsebase/preprocess/preprocess.cc:1705]: Yeah you're right.

File: src/sparsebase/preprocess/preprocess.cc (hunk 6 above)
Comment (AmroAlJundi) [reply src/sparsebase/preprocess/preprocess.cc:1700]: Facts.

Comment (AmroAlJundi) [reply src/sparsebase/preprocess/preprocess.cc:1700]: Good point.

File: src/sparsebase/preprocess/preprocess.cc (hunk 5 above)
Comment (AmroAlJundi) [reply src/sparsebase/preprocess/preprocess.cc]: Yes, you're right.

**Review by Atahanak — APPROVED** [review]

**Pull request merged**